pkg/driver_cryptocell_310/psa_cryptocell_310/cipher_chacha20.c

```diff
@@ -114,46 +116,53 @@ psa_status_t psa_cipher_chacha20_update(psa_cipher_chacha20_ctx_t *ctx,
     DEBUG("Peripheral ChaCha20 Cipher update");
     psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
 
-    if (output_size < input_length)
+    if (output_size < (((ctx->buffer_length + input_length) >> 6) << 6))
     {
         return PSA_ERROR_BUFFER_TOO_SMALL;
     }
 
-    if (!ctx->is_finished)
+    size_t input_index = 0;
+    size_t output_index = 0;
+    CRYSError_t periph_status;
+    // check if buffer contains data
+    cryptocell_310_enable();
+    if (ctx->buffer_length)
     {
-        const uint32_t blocks_length = (input_length / CRYS_CHACHA_BLOCK_SIZE_IN_BYTES) * CRYS_CHACHA_BLOCK_SIZE_IN_BYTES;
-        CRYSError_t periph_status;
-        /* xcrypt full blocks */
-        if (blocks_length) {
-            cryptocell_310_enable();
-            periph_status = CRYS_CHACHA_Block(  &ctx->ctx.ctx,
-                                                            (uint8_t*)input,
-                                                            blocks_length,
-                                                            output);
-            
-            if (periph_status != CRYS_OK) {
-                return CRYS_to_psa_error(periph_status);
-            }
-        }
+        // if yes, fill buffer up and update this block of data
+        input_index = sizeof(ctx->buffer) - ctx->buffer_length;
+        memcpy(&ctx->buffer[ctx->buffer_length], input, input_index);
```

Comment:
```suggestion
        input_size = sizeof(ctx->buffer) - ctx->buffer_length;
        memcpy(&ctx->buffer[ctx->buffer_length], input, input_size);
```

?

Reply:
It is not really a size, but the startindex after we have run update on the buffer